File: android/app/src/main/java/io/blazewallet/breez/BreezBridgePlugin.java
package io.blazewallet.breez;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import breez_sdk.BreezServices;
import breez_sdk.ConnectRequest;
import breez_sdk.Config;
import breez_sdk.NodeConfig;
import breez_sdk.GreenlightNodeConfig;
import breez_sdk.EventListener;
import breez_sdk.ReceivePaymentRequest;
import breez_sdk.SendPaymentRequest;

@CapacitorPlugin(name = "BreezBridge")
public class BreezBridgePlugin extends Plugin {
    
    private BreezServices breezServices;
    
    @PluginMethod
    public void connect(PluginCall call) {
        try {
            String certificate = call.getString("certificate");
            
            if (certificate == null) {
                call.reject("Certificate is required");
                return;
            }
            
            // Setup Greenlight config
            GreenlightNodeConfig greenlightConfig = new GreenlightNodeConfig(
                null, // partner credentials (optional)
                certificate.getBytes()
            );
            
            NodeConfig nodeConfig = NodeConfig.greenlight(greenlightConfig);
            
            Config config = new Config(
                "bitcoin", // network
                System.getProperty("user.home") + "/.breez", // working dir
                nodeConfig,
                null // event listener (optional)
            );
            
            ConnectRequest req = new ConnectRequest(config, System.getProperty("user.home") + "/.breez/seed");
            
            breezServices = BreezServices.connect(req, new EventListener() {
                @Override
                public void onEvent(Event e) {
                    // Forward events to JS
                    notifyListeners("breezEvent", null);
                }
            });
            
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to connect to Breez: " + e.getMessage());
        }
    }
    
    @PluginMethod
    public void getNodeInfo(PluginCall call) {
        try {
            if (breezServices == null) {
                call.reject("Not connected");
                return;
            }
            
            NodeState nodeInfo = breezServices.nodeInfo();
            
            JSObject ret = new JSObject();
            ret.put("id", nodeInfo.id);
            ret.put("maxPayable", nodeInfo.maxPayableMsat / 1000);
            ret.put("maxReceivable", nodeInfo.maxReceivableMsat / 1000);
            ret.put("channelsBalanceMsat", nodeInfo.channelsBalanceMsat / 1000);
            
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to get node info: " + e.getMessage());
        }
    }
    
    @PluginMethod
    public void createInvoice(PluginCall call) {
        try {
            if (breezServices == null) {
                call.reject("Not connected");
                return;
            }
            
            Long amountSats = call.getLong("amountSats");
            String description = call.getString("description");
            
            if (amountSats == null) {
                call.reject("Amount is required");
                return;
            }
            
            ReceivePaymentRequest req = new ReceivePaymentRequest(
                amountSats * 1000, // Convert to msats
                description != null ? description : ""
            );
            
            ReceivePaymentResponse response = breezServices.receivePayment(req);
            
            JSObject ret = new JSObject();
            ret.put("bolt11", response.lnInvoice.bolt11);
            ret.put("paymentHash", response.lnInvoice.paymentHash);
            
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to create invoice: " + e.getMessage());
        }
    }
    
    @PluginMethod
    public void payInvoice(PluginCall call) {
        try {
            if (breezServices == null) {
                call.reject("Not connected");
                return;
            }
            
            String bolt11 = call.getString("bolt11");
            
            if (bolt11 == null) {
                call.reject("Invoice is required");
                return;
            }
            
            SendPaymentRequest req = new SendPaymentRequest(bolt11, null);
            SendPaymentResponse response = breezServices.sendPayment(req);
            
            JSObject ret = new JSObject();
            ret.put("paymentHash", response.payment.id);
            ret.put("amountSats", response.payment.amountMsat / 1000);
            
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to pay invoice: " + e.getMessage());
        }
    }
    
    @PluginMethod
    public void disconnect(PluginCall call) {
        try {
            if (breezServices != null) {
                breezServices.disconnect();
                breezServices = null;
            }
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to disconnect: " + e.getMessage());
        }
    }
}

File: ios/App/App/BreezBridgePlugin.swift
import Foundation
import Capacitor
import BreezSDK

@objc(BreezBridgePlugin)
public class BreezBridgePlugin: CAPPlugin {
    
    private var breezServices: BlockingBreezServices?
    
    @objc func connect(_ call: CAPPluginCall) {
        guard let certificate = call.getString("certificate") else {
            call.reject("Certificate is required")
            return
        }
        
        DispatchQueue.global(qos: .background).async {
            do {
                let greenlightConfig = GreenlightNodeConfig(
                    partnerCredentials: nil,
                    inviteCode: Data(certificate.utf8)
                )
                
                let nodeConfig = NodeConfig.greenlight(config: greenlightConfig)
                
                let config = Config(
                    breezserver: "https://bs1.breez.technology",
                    chainnotifierUrl: "https://chainnotifier.breez.technology",
                    mempoolspaceUrl: nil,
                    workingDir: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0].path + "/.breez",
                    network: Network.bitcoin,
                    paymentTimeoutSec: 60,
                    defaultLspId: nil,
                    apiKey: nil,
                    maxfeePercent: 0.5,
                    exemptfee: Satoshi(value: 25),
                    nodeConfig: nodeConfig
                )
                
                let seed = try mnemonicToSeed(phrase: "your mnemonic here") // TODO: Get from secure storage
                
                let connectRequest = ConnectRequest(config: config, seed: seed)
                
                self.breezServices = try connect(req: connectRequest, eventListener: { event in
                    // Forward events to JS
                    self.notifyListeners("breezEvent", data: [:])
                })
                
                call.resolve()
            } catch {
                call.reject("Failed to connect: \(error.localizedDescription)")
            }
        }
    }
    
    @objc func getNodeInfo(_ call: CAPPluginCall) {
        guard let breez = breezServices else {
            call.reject("Not connected")
            return
        }
        
        do {
            let nodeInfo = try breez.nodeInfo()
            
            call.resolve([
                "id": nodeInfo.id,
                "maxPayable": nodeInfo.maxPayableMsat / 1000,
                "maxReceivable": nodeInfo.maxReceivableMsat / 1000,
                "channelsBalanceMsat": nodeInfo.channelsBalanceMsat / 1000
            ])
        } catch {
            call.reject("Failed to get node info: \(error.localizedDescription)")
        }
    }
    
    @objc func createInvoice(_ call: CAPPluginCall) {
        guard let breez = breezServices else {
            call.reject("Not connected")
            return
        }
        
        guard let amountSats = call.getInt("amountSats") else {
            call.reject("Amount is required")
            return
        }
        
        let description = call.getString("description") ?? ""
        
        do {
            let req = ReceivePaymentRequest(
                amountMsat: UInt64(amountSats * 1000),
                description: description,
                preimage: nil,
                openingFeeParams: nil,
                useDescriptionHash: nil,
                expiry: nil,
                cltv: nil
            )
            
            let response = try breez.receivePayment(req: req)
            
            call.resolve([
                "bolt11": response.lnInvoice.bolt11,
                "paymentHash": response.lnInvoice.paymentHash
            ])
        } catch {
            call.reject("Failed to create invoice: \(error.localizedDescription)")
        }
    }
    
    @objc func payInvoice(_ call: CAPPluginCall) {
        guard let breez = breezServices else {
            call.reject("Not connected")
            return
        }
        
        guard let bolt11 = call.getString("bolt11") else {
            call.reject("Invoice is required")
            return
        }
        
        do {
            let req = SendPaymentRequest(bolt11: bolt11, amountMsat: nil)
            let response = try breez.sendPayment(req: req)
            
            call.resolve([
                "paymentHash": response.payment.id,
                "amountSats": response.payment.amountMsat / 1000
            ])
        } catch {
            call.reject("Failed to pay invoice: \(error.localizedDescription)")
        }
    }
    
    @objc func disconnect(_ call: CAPPluginCall) {
        if breezServices != nil {
            do {
                try breezServices?.disconnect()
                breezServices = nil
                call.resolve()
            } catch {
                call.reject("Failed to disconnect: \(error.localizedDescription)")
            }
        } else {
            call.resolve()
        }
    }
}

